File: 5-longest-palindromic-substring/longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int maxLength = 0;
        int globalLeft = -1;
        int globalRight = -1;

        for (int i = 0; i < s.length(); i++) {
            int left = i, right = i;
            while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
                if (right - left + 1 > maxLength) {
                    maxLength = right - left + 1;
                    globalLeft = left;
                    globalRight = right;
                }
                left--;
                right++;
            }

            left = i; right = i + 1;
            while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
                if (right - left + 1 > maxLength) {
                    maxLength = right - left + 1;
                    globalLeft = left;
                    globalRight = right;
                }
                left--;
                right++;
            }
        }
        return s.substring(globalLeft, globalRight + 1);
    }
}

File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        best = ""
        longest = 0
        for i in range(len(s)):
            length = 1
            j = k = i
            while j > 0 and k < len(s) - 1 and s[j - 1] == s[k + 1]:
                length += 2
                j -= 1
                k += 1
            if length > longest:
                longest = length
                best = s[j:k+1]
        for i in range(len(s) - 1):
            length = 0
            j = i
            k = i + 1
            while j >= 0 and k < len(s) and s[j] == s[k]:
                length += 2
                j -= 1
                k += 1
            if length > longest:
                longest = length
                best = s[j+1:k]
        return best